config/runtimes/vllm-multinode-template.yaml

```diff
@@ -0,0 +1,256 @@
+            - name: HF_HUB_CACHE
+              value: /tmp
+          resources:
+            limits:
+              cpu: "16"
+              memory: 48Gi
+              nvidia.com/gpu: "1"
+            requests:
+              cpu: "8"
+              nvidia.com/gpu: "1"
+          volumeMounts:
+            - name: shm
+              mountPath: /dev/shm
+          livenessProbe:
+            failureThreshold: 3
+            periodSeconds: 30
+            successThreshold: 1
+            timeoutSeconds: 8
+            exec:
+              command:
+                - bash
+                - -c
+                - |
+                  # Check if the registered ray nodes count is the same as PIPELINE_PARALLEL_SIZE
+                  gpu_status=$(ray status | grep GPU)
+                  if [[ -z $gpu_status ]]; then
+                    echo "$1: GPU does not exist"
+                    exit 1
+                  fi
+
+                  used_gpu=$(echo "$gpu_status" | awk '{print $1}' | cut -d'/' -f1)
+                  reserved_gpu=$(echo "$gpu_status" | awk '{print $1}' | cut -d'/' -f2)
+
+                  # Determine health status based on GPU usage
+                  if [[ "$used_gpu" != "$reserved_gpu" ]]; then
+                    echo "$1: Unhealthy - Used: $used_gpu, Reserved: $reserved_gpu"
+                    exit 1
+                  fi
+          readinessProbe:
+            failureThreshold: 2
+            periodSeconds: 10
+            successThreshold: 1
+            timeoutSeconds: 5
+            exec:
+              command:
+                - bash
+                - -c
+                - |
+                  # Check if the registered nodes count matches PIPELINE_PARALLEL_SIZE
+                  registered_node_count=$(ray status | grep -c node_)
+                  if [[ $registered_node_count -ne "$PIPELINE_PARALLEL_SIZE" ]]; then
+                    echo "Readiness Probe: Unhealthy - Registered nodes count ($registered_node_count) does not match PIPELINE_PARALLEL_SIZE ($PIPELINE_PARALLEL_SIZE)."
+                    exit 1
+                  fi
+
+                  # Check if the registered ray nodes count is the same as PIPELINE_PARALLEL_SIZE
+                  gpu_status=$(ray status | grep GPU)
+                  if [[ -z $gpu_status ]]; then
+                    echo "$1: GPU does not exist"
+                    exit 1
+                  fi
+
+                  used_gpu=$(echo "$gpu_status" | awk '{print $1}' | cut -d'/' -f1)
+                  reserved_gpu=$(echo "$gpu_status" | awk '{print $1}' | cut -d'/' -f2)
+
+                  # Determine health status based on GPU usage
+                  if [[ "$used_gpu" != "$reserved_gpu" ]]; then
+                    echo "$1: Unhealthy - Used: $used_gpu, Reserved: $reserved_gpu"
+                    exit 1
+                  fi
+
+                  # Check model health
+                  curl --silent --max-time 5 --fail-with-body http://localhost:8080/health
```

Comment:
Probes will likely encounter problems with large models, some models could take several minutes to respond

Reply:
right. it took around 12 minutes to load a 405B model. Therefore, I will be adding a health check (healthz) for the runtime during startup.

So I will add model healthz check in the StartupProbe.

*Startup Probe:*
- Purpose: To verify that ray is up and that the runtime returns the model as ready.
- Timeout: It will check for up to 20 minutes.

*Readiness Probe:*
- Purpose: After the Startup Probe succeeds, this probe checks if the environment is ready to serve requests.
- Checks Include:
  - Verifying the count of ray nodes and their readiness.
  - Checking the model's health status.
- Timeout: 30 secs

*Liveness Probe:*
- Purpose: To continuously monitor the running state of the ray nodes.

**Timeline Summary**
- Startup Probe: Runs for up to 20 minutes to ensure the model is loaded and ready.
- Readiness Probe: Activated after the Startup Probe passes; it checks the health of the nodes and the model.
- Liveness Probe: Monitors the health of the ray nodes throughout its lifecycle.

@Xaenalt wdyt?

Reply:
Yeah, it's a good idea to use Startup and Readiness/Liveness probes. I think they're finally supported now:

https://issues.redhat.com/browse/SRVKS-1186

Reply:
multi node feature will be only supported with `rawDeployment`